src/features/ajax/aggregate/index.js

```diff
@@ -15,251 +14,183 @@ import { SUPPORTABILITY_METRIC_CHANNEL } from '../../metrics/constants'
+    event.gql = params.gql = parseGQL({
+      body: ctx.body,
+      query: ctx.parsedOrigin?.search
+    })
+    if (event.gql) handle(SUPPORTABILITY_METRIC_CHANNEL, ['Ajax/Events/GraphQL/Bytes-Added', stringify(event.gql).length], undefined, FEATURE_NAMES.metrics, this.ee)
+
+    const softNavInUse = Boolean(getNREUMInitializedAgent(this.agentIdentifier)?.features?.[FEATURE_NAMES.softNav])
+    if (softNavInUse) { // For newer soft nav (when running), pass the event to it for evaluation -- either part of an interaction or is given back
+      handle('ajax', [event], undefined, FEATURE_NAMES.softNav, this.ee)
+    } else if (ctx.spaNode) { // For old spa (when running), if the ajax happened inside an interaction, hold it until the interaction finishes
+      const interactionId = ctx.spaNode.interaction.id
+      this.spaAjaxEvents[interactionId] = this.spaAjaxEvents[interactionId] || []
+      this.spaAjaxEvents[interactionId].push(event)
+    } else {
+      this.ajaxEvents.push(event)
+    }
+  }
 
-      if (ajaxEvents.length === 0) {
-        return null
-      }
+  prepareHarvest (options) {
+    options = options || {}
+    if (this.ajaxEvents.length === 0) return null
 
-      var payload = getPayload(ajaxEvents, options.maxPayloadSize || MAX_PAYLOAD_SIZE)
+    const payload = this.#getPayload(this.ajaxEvents)
+    const payloadObjs = []
 
-      var payloadObjs = []
-      for (var i = 0; i < payload.length; i++) {
-        payloadObjs.push({ body: { e: payload[i] } })
-      }
+    for (let i = 0; i < payload.length; i++) payloadObjs.push({ body: { e: payload[i] } })
 
-      if (options.retry) {
-        sentAjaxEvents = ajaxEvents.slice()
-      }
+    if (options.retry) this.sentAjaxEvents = this.ajaxEvents
+    this.ajaxEvents = []
 
-      ajaxEvents = []
+    return payloadObjs
+  }
 
-      return payloadObjs
+  onEventsHarvestFinished (result) {
+    if (result.retry && this.sentAjaxEvents.length > 0) {
+      this.ajaxEvents.unshift(...this.sentAjaxEvents)
+      this.sentAjaxEvents = []
     }
+  }
 
-    function getPayload (events, maxPayloadSize, chunks) {
-      chunks = chunks || 1
-      var payload = []
-      var chunkSize = events.length / chunks
-      var eventChunks = splitChunks(events, chunkSize)
-      var tooBig = false
-      for (var i = 0; i < eventChunks.length; i++) {
-        var currentChunk = eventChunks[i]
-        if (currentChunk.tooBig(maxPayloadSize)) {
-          if (currentChunk.events.length !== 1) {
-            /* if it is too big BUT it isnt length 1, we can split it down again,
-             else we just want to NOT push it into payload
-             because if it's length 1 and still too big for the maxPayloadSize
-             it cant get any smaller and we dont want to recurse forever */
-            tooBig = true
-            break
-          }
-        } else {
-          payload.push(currentChunk.payload)
+  #getPayload (events, numberOfChunks) {
```

Comment:
Again since its on topic, what is the designation for when a method should be private vs public?

Reply:
`getPayload` should not be an accessible function outside the Aggregate instance (as there's no use case for it alone at the moment; it's called by `storeXhr`). If it weren't for tests, maybe neither should `prepareHarvest`. The reason why it's not a separate function is because it's inherently tied to this class and isn't re-usable by other classes. It could exist as a pure fn though if we pass `this` to it strictly speaking, as this is a OOP principle more than a technical restriction.